### PR TITLE
[PS-1341] Add folder to item view

### PR DESCRIPTION
**Pull request by patrickhlauke**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Currently, the item view does not give any indication if the item is in a folder or not. This PR adds an extra"Folder: ..." label/value to the item info.

## Code changes

- **apps/browser/src/popup/vault/view.component.html** and **apps/desktop/src/app/vault/view.component.html**: add "Folder:..." label/value box to the view
- **apps/browser/src/popup/vault/view.component.ts** and **apps/desktop/src/app/vault/view.component.ts**: add `FolderView` to the definition
- **libs/angular/src/components/view.component.ts**: add code to turn `folderId` into actual reference to the specific folder the cipher is in, if any (with thanks to @Hinton for pushing me in the right direction)

## Screenshots
A test item in the popup's item view, showing the new 'Folder: ...' field after the web addresses

A test item in the deskop's item view, showing the new 'Folder: ...' field after the web addresses

## Before you submit

```
- [x] I have checked for **linting** errors (`npm run lint`) (required)
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)
```

### CODE REVIEW

**Comment by patrickhlauke** [comment]
@djsmith85 my angular skills/understanding of how the codebase is organised here fail me ... I can grab the folder id, but not sure how to now grab the actual text/name of the folder. something involving `folders$`, but not quite sure how to grab/populate that. any chance you could point me in the right direction, or maybe get somebody to collab with me on this?

**Comment by patrickhlauke** [comment]
also pinging @Hinton just in case he knows :)

**Comment by Hinton** [comment]
@patrickhlauke you should be able to grab the folder from the id using something like
```ts
const folder (await firstValueFrom(this.folderService.folderView$)).find(f => f.id == <id>);
```

**Comment by patrickhlauke** [comment]
@hinton sadly, my lack of angular knowledge now means i have zero idea where to slot that in...possibly bitten off more than i can chew here.

**Comment by Hinton** [comment]
Hmm, one solution would be to add a `folder` field to `libs/angular/src/components/view.component.ts` called `folder`, and fetching it in `load` after `cipher.decrypt`.

```
this.folder (await firstValueFrom(this.folderService.folderView$)).find(f => f.id == <id>);
```

That would also allow you to use it in the other clients. The folderService would need to be added as a constructor argument.

**Comment by patrickhlauke** [comment]
incidentally, while this is not ready yet (not found the time to actually work out how to do this last bit of getting the actual folder name), wondering if @danielleflinn thinks this stands a chance to make it into the product as well (as it's adjacent to something like https://github.com/bitwarden/clients/pull/3181)?

**Comment by sukhleenb** [comment]
Hey Patrick! We do think including the folder name will be valuable data to display on this page. Since we’re looking into adding the Created date (# [3181](https://github.com/bitwarden/clients/pull/3181)), we’re leaning towards displaying the folder name in a slightly different pattern.  

Let me know if you’d like to change your PR to reflect the changes shown in the screenshot (folder name form field UI beneath the website field). If the item does not exist in a folder, we just wouldn't show that field. 

Screen Shot 2022-09-08 at 12 13 46 PM

**Comment by patrickhlauke** [comment]
> Created date (# https://github.com/bitwarden/clients/pull/3181), we’re leaning towards displaying the folder name in a slightly different pattern.

>
> Let me know if you’d like to change your PR to reflect the changes shown in the screenshot (folder name form field UI beneath the website field).

yup, looks good. was initially aiming at the most minimal UI change, but if you're happy with the more obvious one shown in that screenshot, I'll go for something like it (once I worked out the missing bit of how to get the actual human-readable folder name)

**Comment by patrickhlauke** [comment]
with thanks to @Hinton for putting me on the right track (though it still needed a lot of trial and error on my part), this is now ready for review :)

**Comment by patrickhlauke** [comment]
if https://github.com/bitwarden/clients/pull/3321 gets merged first, this will also need to have the drag-n-drop bit added

**Comment by patrickhlauke** [comment]
> if #3321 gets merged first, this will also need to have the drag-n-drop bit added

done

**Comment by patrickhlauke** [comment]
Updated to use `<label>` / `<input type="text" readonly>` (in anticipation of https://github.com/bitwarden/clients/pull/3485)

File: libs/angular/src/components/view.component.ts [hunk 1]
```diff
@@ -112,6 +117,10 @@ export class ViewComponent implements OnDestroy, OnInit {
     this.showPremiumRequiredTotp =
       this.cipher.login.totp && !this.canAccessPremium && !this.cipher.organizationUseTotp;
 
+    this.folder = await (
+      await firstValueFrom(this.folderService.folderViews$)
+    ).find((f) => f.id == this.cipher.folderId);
+
```
Comment (djsmith85) [inline libs/angular/src/components/view.component.ts:123]: We should add a if-condition to only fetch the folderName when the folderId is present/filled.

**Review by djsmith85 — CHANGES_REQUESTED** [review]
@patrickhlauke Thank you for your contribution. Just some minor things I've noticed.

**Comment by patrickhlauke** [comment]
Thanks @djsmith85 ... I'll have a look at all the above comments in a moment :+1:

File: libs/angular/src/components/view.component.ts (hunk 1 above)
Comment (patrickhlauke) [reply libs/angular/src/components/view.component.ts]: Done

File: apps/browser/src/popup/vault/view.component.html [hunk 2]
```diff
@@ -468,6 +468,20 @@ <h2 class="box-header">
       </div>
     </div>
   </div>
+  <div class="box" *ngIf="cipher.folderId">
+    <div class="box-content">
+      <div class="box-content-row">
+        <label
+          for="folder"
```
[the remark this answers is not in the document]
Comment (patrickhlauke) [reply apps/browser/src/popup/vault/view.component.html]: Oops. fixed

File: apps/browser/src/popup/vault/view.component.html [hunk 3]
```diff
@@ -468,6 +468,20 @@ <h2 class="box-header">
...
+          class="draggable"
+          draggable="true"
+          (dragstart)="setTextDataOnDrag($event, folder.name)"
+          >{{ "folder" | i18n }}</label
+        >
+        <input id="FolderId" type="text" name="folderId" [value]="folder.name" readonly />
```
[the remark this answers is not in the document]
Comment (patrickhlauke) [reply apps/browser/src/popup/vault/view.component.html]: Done

File: apps/browser/src/popup/vault/view.component.html [hunk 4]
```diff
@@ -468,6 +468,20 @@ <h2 class="box-header">
...
+    <div class="box-content">
+      <div class="box-content-row">
+        <label
+          for="folderName"
+          class="draggable"
+          draggable="true"
+          (dragstart)="setTextDataOnDrag($event, folder.name)"
+          >{{ "folder" | i18n }}</label
+        >
+        <input id="folderName" type="text" name="folderId" [value]="folder.name" readonly />
```
Comment (djsmith85) [inline apps/browser/src/popup/vault/view.component.html:481]: @patrickhlauke Thanks for updating the PR. I think we should still change the name-attribute. The element id is looking good.

**Review by djsmith85 — CHANGES_REQUESTED** [review]

File: apps/browser/src/popup/vault/view.component.html (hunk 4 above)
Comment (patrickhlauke) [reply apps/browser/src/popup/vault/view.component.html]: Done

**Review by djsmith85 — APPROVED** [review]
@patrickhlauke This is looking good. Thanks again for you contributions.

I'll pass this on to QA and once they've given their approval I'll get this merged into master.

**Comment by djsmith85** [comment]
@patrickhlauke I've received approval from QA and this is ready to get merged. Thank you for your work on this.

**Review by djsmith85 — APPROVED** [review]

**Pull request merged**